### PR TITLE
Suggesting changes to #338

### PR DESCRIPTION
**Pull request by NicolaCourtier**
# Description

Demonstrating alternative changes to implement the additions in #338. The main difference is how the `sigma` parameter is treated in the `GaussianLogLikelihood` cost function. Here, it is implemented as an additional PyBOP `Parameter`.

## Issue reference
Towards fixing #257.

## Review
Before you mark your PR as ready for review, please ensure that you've considered the following:
- Updated the [CHANGELOG.md](https://github.com/pybop-team/PyBOP/blob/develop/CHANGELOG.md) in reverse chronological order (newest at the top) with a concise description of the changes, including the PR number.
- Noted any breaking changes, including details on how it might impact existing functionality.

## Type of change
- [ ] New Feature: A non-breaking change that adds new functionality.
- [ ] Optimization: A code change that improves performance.
- [ ] Examples: A change to existing or additional examples.
- [ ] Bug Fix: A non-breaking change that addresses an issue.
- [ ] Documentation: Updates to documentation or new documentation for new features.
- [ ] Refactoring: Non-functional changes that improve the codebase.
- [ ] Style: Non-functional changes related to code style (formatting, naming, etc).
- [ ] Testing: Additional tests to improve coverage or confirm functionality.
- [ ] Other: (Insert description of change)

# Key checklist:

- [ ] No style issues: `$ pre-commit run` (or `$ nox -s pre-commit`) (see [CONTRIBUTING.md](https://github.com/pybop-team/PyBOP/blob/develop/CONTRIBUTING.md#installing-and-using-pre-commit) for how to set this up to run automatically when committing locally, in just two lines of code)
- [ ] All unit tests pass: `$ nox -s tests`
- [ ] The documentation builds: `$ nox -s doctest`

You can run integration tests, unit tests, and doctests together at once, using `$ nox -s quick`.

## Further checks:
- [ ] Code is well-commented, especially in complex or unclear areas.
- [ ] Added tests that prove my fix is effective or that my feature works.
- [ ] Checked that coverage remains or improves, and added tests if necessary to maintain or increase coverage.

Thank you for contributing to our project! Your efforts help us to deliver great software.

### CODE REVIEW

**Comment by BradyPlanden** [comment]
Hi @NicolaCourtier - thanks for this alternative implementation! I went down a similar path before stopping during the implementation of #338, I came to the conclusion that adding the `sigma` hyper into the constructed `Parameters` requires quite a bit of patching. Having the parameters constructed early in the life of an optimisation task and then building the `x0`, `sigma0`, and `boundaries` from them provides a clean separation between parameters -> model and cost -> optim. At the point of the cost and optim construction, we only need to worry about / modify `x0`, `sigma0`, and `boundaries` attributes, which is very easy to understand and review. I'd like to continue with the `x0`, `sigma0`, and `boundaries` implementation in #338, as I think it's concise and provides simplicity.

However, I'm keen to integrate your general improvements into #338. Specifically, your updates to [check_sigma0](https://github.com/pybop-team/PyBOP/pull/352/files#diff-bc69ba808ed48e174069336685bcf53b9fac3313425a19278b258dc474760776R87-R92), [plot_2d](https://github.com/pybop-team/PyBOP/pull/352/files#diff-e1c12edf3c7afaebb7ecda90f4e5b017929b1ee73617f64ef1b401130c95039eR58-R68) and the [CMAES](https://github.com/pybop-team/PyBOP/pull/352/files#diff-3f05a644a8efd509c952c823f37e77efcd198bbea3cd13241945caa30cb10e9dR237) check look great! Also, all of the docstrings changes I had missed 😄. If you'd like to commit those changes directly to that branch, that would be amazing, otherwise I'm happy to add them during review.

Have a great weekend!

**Comment by NicolaCourtier** [comment]
Thanks for the comment @BradyPlanden, I'm just getting the coverage up on this branch and then it would be great to discuss the different implementations tomorrow, if you have time.

**Comment by NicolaCourtier** [comment]
There will be some conflicts between the changes to `parameters.as_dict()` in this branch and the change from parameters to inputs in #358. Merging #358 (when ready) may help to clean up the implementation in this branch.

**Comment by NicolaCourtier** [comment]
I have merged #359 into this one and resolved the conflicts. Merging the same into #338 will significantly reduce the number of files changed.

File: pybop/costs/_likelihoods.py [hunk 1]
```diff
@@ -101,20 +113,46 @@ def __init__(
...
+                )
+            else:
+                raise TypeError(
+                    "Expected sigma0 to contain Parameter objects or numeric values. "
+                    + f"Received {type(s0)}"
+                )
+        self.parameters.join(self.sigma)
+
+        if dsigma_scale is None:
+            self._dsigma_scale = sigma0
```
Comment (BradyPlanden) [inline pybop/costs/_likelihoods.py:150]: Change in #338 to `self._dsgima_scale == 1.0`

File: pybop/costs/_likelihoods.py [hunk 2]
```diff
@@ -129,43 +167,29 @@ def dsigma_scale(self, new_value):
...
-        sigma = np.asarray(x[-self.n_outputs :])
+        self.parameters.update(values=list(inputs.values()))
+
+        sigma = self.sigma.current_value()
         if np.any(sigma <= 0):
             return -np.inf
 
-        y = self.problem.evaluate(x[: -self.n_outputs])
+        problem_inputs = self.problem.parameters.as_dict()
+        y = self.problem.evaluate(problem_inputs)
```
Comment (BradyPlanden) [inline pybop/costs/_likelihoods.py:192]: Change in #338 to: 
```suggestion
        y = self.problem.evaluate(self.problem.parameters.as_dict())
```

File: pybop/costs/_likelihoods.py [hunk 3]
```diff
@@ -184,34 +208,39 @@ def _evaluate(self, x: np.ndarray, grad: Union[None, np.ndarray] = None) -> floa
...
-        sigma = np.asarray(x[-self.n_outputs :])
+        self.parameters.update(values=list(inputs.values()))
+
+        sigma = self.sigma.current_value()
         if np.any(sigma <= 0):
             return -np.inf, -self._dl
 
-        y, dy = self.problem.evaluateS1(x[: -self.n_outputs])
+        problem_inputs = self.problem.parameters.as_dict()
+        y, dy = self.problem.evaluateS1(problem_inputs)
```
Comment (BradyPlanden) [inline pybop/costs/_likelihoods.py:232]: change in #338 to:
```suggestion
        y, dy = self.problem.evaluateS1(self.problem.parameters.as_dict())
```

File: pybop/costs/_likelihoods.py [hunk 4]
```diff
@@ -184,34 +208,39 @@ def _evaluate(self, x: np.ndarray, grad: Union[None, np.ndarray] = None) -> floa
...
 
+        likelihood = self._evaluate(inputs)
+
         r = np.array([self._target[signal] - y[signal] for signal in self.signal])
-        likelihood = self._evaluate(x)
-        dl = np.sum((sigma ** (-2.0) * np.sum((r * dy.T), axis=2)), axis=1)
+        dl = np.sum((np.sum((r * dy.T), axis=2) / (sigma**2)), axis=1)
         dsigma = (
-            -self.n_time_data / sigma + sigma ** (-3.0) * np.sum(r**2, axis=1)
+            -self.n_time_data / sigma + np.sum(r**2, axis=1) / (sigma**3)
```
Comment (BradyPlanden) [inline pybop/costs/_likelihoods.py:243]: change in #338 to:
```suggestion
            -self.n_time_data / sigma + np.sum(r**2, axis=1) / (sigma**3.0)
```

File: pybop/costs/_likelihoods.py [hunk 5]
```diff
@@ -184,34 +208,39 @@ def _evaluate(self, x: np.ndarray, grad: Union[None, np.ndarray] = None) -> floa
...
             len(y.get(key, [])) != len(self._target.get(key, [])) for key in self.signal
         ):
             return -np.inf, -self._dl
 
+        likelihood = self._evaluate(inputs)
+
         r = np.array([self._target[signal] - y[signal] for signal in self.signal])
-        likelihood = self._evaluate(x)
-        dl = np.sum((sigma ** (-2.0) * np.sum((r * dy.T), axis=2)), axis=1)
+        dl = np.sum((np.sum((r * dy.T), axis=2) / (sigma**2)), axis=1)
```
Comment (BradyPlanden) [inline pybop/costs/_likelihoods.py:241]: change in #338 to:
```suggestion
        dl = np.sum((np.sum((r * dy.T), axis=2) / (sigma**2.0)), axis=1)
```

File: pybop/costs/_likelihoods.py [hunk 6]
```diff
@@ -296,10 +325,21 @@ def _evaluateS1(self, x):
...
+        delta = 1e-3
+        dl_prior_approx = [
+            (
+                param.prior.logpdf(inputs[param.name] * (1 + delta))
+                - param.prior.logpdf(inputs[param.name] * (1 - delta))
+            )
+            / (2 * delta * inputs[param.name] + np.finfo(float).eps)
+            for param in self.problem.parameters
+        ]
+
```
Comment (BradyPlanden) [inline pybop/costs/_likelihoods.py:343]: Change in #340, or split the prior update for evaluate and evaluate_S1 into separate PR.

File: pybop/models/base_model.py [hunk 7]
```diff
@@ -470,10 +468,9 @@ def predict(
 
         Parameters
         ----------
-        inputs : dict or array-like, optional
-            Input parameters for the simulation. If the input is array-like, it is converted
-            to a dictionary using the model's fitting keys. Defaults to None, indicating
-            that the default parameters should be used.
+        inputs : Inputse, optional
```
Comment (BradyPlanden) [inline pybop/models/base_model.py:471]: Update in #338: 
```suggestion
        inputs : Inputs, optional
```

File: tests/integration/test_spm_parameterisations.py [hunk 8]
```diff
@@ -72,7 +72,7 @@ def spm_costs(self, model, parameters, cost_class, init_soc):
         if cost_class in [pybop.GaussianLogLikelihoodKnownSigma]:
             return cost_class(problem, sigma0=0.002)
         elif cost_class in [pybop.GaussianLogLikelihood]:
-            return cost_class(problem)
+            return cost_class(problem, sigma0=0.002 * 3)
```
Comment (BradyPlanden) [inline tests/integration/test_spm_parameterisations.py:75]: Update in #338 to:
```suggestion
            return cost_class(problem, sigma0=0.002 * 3) # Initial sigma0 guess
```

File: tests/unit/test_likelihoods.py [hunk 9]
```diff
@@ -127,7 +132,25 @@ def test_gaussian_log_likelihood(self, one_signal_problem):
         grad_result, grad_likelihood = likelihood.evaluateS1(np.array([0.5, 0.5]))
         assert isinstance(result, float)
         np.testing.assert_allclose(result, grad_result, atol=1e-5)
-        assert np.all(grad_likelihood <= 0)
+        assert grad_likelihood[0] <= 0  # TEMPORARY WORKAROUND
```
Comment (BradyPlanden) [inline tests/unit/test_likelihoods.py:135]: Update in #338

**Review by BradyPlanden — APPROVED** [review]
Thanks @NicolaCourtier, I've added a few comments for me to address in #338. Will merge now.

**Pull request merged**